doc/source/ray-air/check-ingest.rst

```diff
@@ -3,444 +3,171 @@
-convert the provided :py:class:`~ray.data.Dataset` into a :py:class:`~ray.data.dataset_pipeline.DatasetPipeline` with the specified configurations.
+Custom data config (advanced)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-See the :ref:`Enabling Streaming Ingest <air-streaming-ingest>` and :ref:`Shuffling Data <air-shuffle>` sections for full examples.
+For use cases not covered by the default config class, you can also fully customize exactly how your input datasets are splitted. To do this, you need to define a custom ``DataConfig`` class (DeveloperAPI). The ``DataConfig`` class is responsible for that shared setup and splitting of data across nodes.
 
-How do I shard validation and test datasets?
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+.. code:: python
 
-By default only the `"train"` Dataset is sharded. To also shard validation and test datasets, you can configure the ``dataset_config``
-that is passed to your ``Trainer``.
-See the :ref:`Splitting Auxiliary Datasets <air-splitting-aux-datasets>` section for a full example.
+    train_ds = ...   # As before.
+
+    # Note this example class is doing the same thing as
+    # the basic DataConfig impl included with Train.
+    class MyCustomDataConfig(train.DataConfig):
+       def get_shards(
+            self,
+            datasets: Dict[str, Dataset],
+            world_size: int,
+            worker_node_ids: List[NodeIdStr],
+            **kwargs):
+
+          # Configure Ray Data for ingest 
+          ctx = ray.data.DataContext.get_current()
+          ctx.execution_options = ctx.ingest_default_options()
+
+          # Split the stream into shards.
+          iterator_shards = datasets["train"].streaming_split(
+               world_size,
+               equal=True,
+               locality_hints=worker_node_ids)
+
+          # Create shards for each worker and return.
+          outputs = {}
+          for rank in range(world_size):
+              outputs[rank] = {"train": iterator_shards[rank]}
+          self.output_splits = outputs
+
+       def get_shard(self, rank: int, name: str) -> DataIterator:
+          assert name == "train", "not implemented"
+          return self.output_splits[rank][name]
+
+    def train_loop_per_worker():
+        # Nothing here changes.
+        ...
+
+    my_trainer = TorchTrainer(
+        train_loop_per_worker,
+        scaling_config=ScalingConfig(num_workers=2),
+        datasets={"train": train_ds},
+        dataset_config=MyCustomDataConfig(),
+    )
+    my_trainer.fit()
+
+
+What do you need to know about this ``DataConfig`` class?
+
+* It must be serializable, since it will be copied from the driver script to the driving actor of the Trainer.
+* Its ``configure`` method is called on the main actor of the Trainer group to setup data ingest and iterators.
+* Train will get data shards for each worker by calling ``get_shard`` on the instance after setup is called.
```

Comment:
Is this already implemented? I didn't find this `get_shard` method in #35236

Reply:
Let me move to doc_code.